core/unit_test/tools/TestBuiltinTuners.cpp

```diff
@@ -13,84 +13,162 @@
+  auto end_point = new_team_tuner.get_point(0.9, 0.0, 0.0);
+  EXPECT_EQ(std::get<0>(end_point), 5);
+  for (int x = 0; x < 10000; ++x) {
+    auto config                 = new_team_tuner.begin();
+    [[maybe_unused]] int option = std::get<0>(config);
+    int team                    = std::get<1>(config);
+    int vector                  = std::get<2>(config);
+    Kokkos::parallel_for("mdrange",
+                         Kokkos::TeamPolicy<ExecSpace>(1, team, vector),
+                         TestTeamFunctor{});
+    new_team_tuner.end();
+  }
+  end_context(kernel_context);
+}
+
+void validate_tile_sizes(const std::vector<int>& cont,
+                         std::array<int, 6> current_tile,
+                         const std::array<int, 3>& hw_tile_limits,
+                         int current_rank, const int policy_rank) {
+  namespace Impl = Kokkos::Tools::Experimental::Impl;
+  for (const auto& size : cont) {
+    current_tile[current_rank] = size;
+    EXPECT_TRUE(Impl::valid_tile(hw_tile_limits, current_tile, policy_rank));
   }
-  Kokkos::finalize();
 }
+
+template <typename Mapped>
+void validate_tile_sizes(std::map<int, Mapped>& cont,
+                         std::array<int, 6> current_tile,
+                         const std::array<int, 3>& hw_tile_limits,
+                         int current_rank, const int policy_rank) {
+  for (auto it = cont.begin(); it != cont.end(); it++) {
+    int key                    = it->first;
+    current_tile[current_rank] = key;
+    validate_tile_sizes(it->second, current_tile, hw_tile_limits,
+                        current_rank + 1, policy_rank);
+  }
+}
+
+// Test helper function that verifies all tiles in the configuration space
+// satisfy hardware constraints.
+template <typename Mapped>
+void validate_tile_sizes(std::map<int, Mapped>& cont,
+                         const std::array<int, 3>& hw_tile_limits,
+                         int policy_rank) {
+  std::array<int, 6> current_tile{1, 1, 1, 1, 1, 1};
+  validate_tile_sizes(cont, current_tile, hw_tile_limits, 0, policy_rank);
+}
+
+template <int test_policy_rank>
+void test_tile_constraints() {
+  using namespace Kokkos::Tools::Experimental::Impl;
+  using SpaceDescription =
+      typename n_dimensional_sparse_structure<int, test_policy_rank>::type;
+
+  std::array<int, 3> hardware_limits{128, 128, 128};
+  SpaceDescription tile_configuration_space;
+  const int max_total_tile_size = 512;
+
+  fill_tile(tile_configuration_space, max_total_tile_size);
+  apply_tiles_constraints(tile_configuration_space, hardware_limits,
+                          test_policy_rank);
+  validate_tile_sizes(tile_configuration_space, hardware_limits,
+                      test_policy_rank);
+}
+
+TEST(kokkosp, constraint_tiles) {
+  namespace Impl = Kokkos::Tools::Experimental::Impl;
+
+  std::array<int, 3> hardware_limits{128, 128, 128};
+  std::array<int, 6> valid_tile{4, 2, 4, 4, 2, 4};
+  std::array<int, 6> valid_tile_max{128, 128, 128, 1, 1};
```

Comment:
Did you mean to only provide 5 numbers (i.e. implicitly zero-initialize the 6th dimension)